File: r-package/morie/R/aitalri.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Map alr coordinates back to a closed composition.
#'
#' \code{ref} is a one-based index into the reconstructed composition of
#' length D = length(y) + 1.
#'
#' Formula: alr^-1(y) = C( exp(y_1), ..., 1 at position ref, ... )
#'
#' @param y alr coordinates, length D-1.
#' @param ref One-based position the reference part is reinstated at.
#' @param total Constant the returned composition sums to.
#' @return List with \code{composition}, \code{ref}, \code{total}, \code{D}.
#' @references Aitchison (1986), The Statistical Analysis of Compositional
#'   Data, Chapter 4. Verified against the reference implementation in the
#'   CRAN package compositions 2.0-9, whose alrInv is clo(exp(cbind(z, 0)))
#'   -- a zero column appended, then closed.
#' @export
Alrinv <- function(y, ref = NULL, total = 1) {
  y <- .t1_vec(y)
  D <- length(y) + 1L
  r <- if (is.null(ref)) D else as.integer(ref)
  if (r < 1L || r > D) stop("ref must be a one-based part index in 1..D")
  full <- append(y, 0, after = r - 1L)
  e <- exp(full)
  s <- sum(e)
  k <- as.numeric(total)
  .t1_result(composition = k * e / s, ref = r, total = k, D = D,
             method = "Inverse additive log-ratio transform")
}

File: src/morie/fn/aitalri.py
# morie.fn -- function file (rootcoder007/morie)
"""Inverse additive log-ratio transform."""

import math

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ['alrinv', 'aitchison_alr_inverse']


def alrinv(y, ref=None, total=1.0):
    """Inverse additive log-ratio transform.

    Formula: alr^-1(y) = C( exp(y_1), ..., exp(y_{D-1}), 1 ) with the 1 inserted at the reference position

    Parameters
    ----------
    y : array-like
        Additive log-ratio coordinates, length D - 1.
    ref : int
        1-based index the reference part is restored to; the default is the last position D.
    total : float
        Constant kappa the closure sums to.

    Returns
    -------
    RichResult
        ``composition``, ``ref``, ``total``, ``D``.

    References
    ----------
    Aitchison, J. (1986), The Statistical Analysis of Compositional Data, Chapman and Hall, is this shelf's primary book and is NOT in the reference library, so it could not be read.  The definitions below were taken instead from Mateu-Figueras, G., Pawlowsky-Glahn, V. and Egozcue, J. J., The normal distribution in some constrained sample spaces, arXiv:0802.2643 (published as SORT 37(1):29-56, 2013), Sect. 4.1, which prints them with equation numbers and attributes them to Aitchison (1982, 1986); that paper was FETCHED and is archived in the reference library with a row in EXTERNAL_SOURCES.md.  Inverting alr puts a 1 back in the reference slot and closes: the reference part is the one whose log-ratio against itself is zero.  ``ref`` must match the ``ref`` used in the forward transform for the round trip to return the original composition.
    """
    y = C.vec(y)
    D = len(y) + 1
    if len(y) == 0:
        raise ValueError("y must be non-empty")
    k = D if ref is None else int(ref)
    if not 1 <= k <= D:
        raise ValueError("ref must be a 1-based part index")
    full = [0.0] * D
    idx = [i for i in range(1, D + 1) if i != k]
    for pos, i in enumerate(idx):
        full[i - 1] = y[pos]
    m = max(full)
    e = [math.exp(v - m) for v in full]
    s = sum(e)
    t = float(total)
    return RichResult(payload={
        "composition": [t * v / s for v in e], "ref": k, "total": t, "D": D,
        "method": "Inverse additive log-ratio transform"})


aitchison_alr_inverse = alrinv


def cheatsheet():
    return 'aitalri: Inverse additive log-ratio transform.'
